### PR TITLE
chain reorg, http endpoint, ds persistence, others

### LINKED ISSUE #4: index/ask: recover and load from datastore
Recover from datastore when spinning up a new cache with saved data.

### LINKED ISSUE #26: api: use persistent datastore
Use Badger as `datastore.TxnDatastore`.

### LINKED ISSUE #28: index: chain reorgs
Decide what is the best strategy to handle chain reorgs when using `ChainNotify`.

### LINKED ISSUE #29: index/miner: persist and load from datastore
Save progress in datastore, and load it when restarting.

### LINKED ISSUE #32: index/slashing: persist and load from datastore
Save progress in datastore, and load it when restarting.

### LINKED ISSUE #61: actions: run at least short tests
Make GH Actions at least run tests that doesn't depend on a Lotus node running.
This is better than the current config that only compiles tests.

### PR DESCRIPTION
**Pull request by jsign**
I think most refered issues give a picture of what was done here.

The central change was using a new API that we PRed to know if two Tipsets are or not from the same fork and act in consequence. Using that, I built a `chainstore` which saves snapshot of state (in general), and is used to get the last saved state (compatible with a newly discovered Tipset) and work from there.

In case of a fork, most probably the last saved state won't be compatible and will go backwards looking for older ones until one is found (or none, and will have to work from genesis).

Now also indexes information is save in persistent datastore.

Apart from that, did a quite severe refactoring, renaming and bugfixes. I think the code is much more structured now and can grow more cleanly since this will keep growing.

Also, I did some work on all test to separate the ones that can be run without a Lotus node from the one that need one. Then, I changed the Github Actions yaml to run the tests that can be run by CI... so now we can at least run real tests than simply compiling them.

Added some http endpoints to quickly have a JSON formatted output of current indexes information.

A bunch of other things were more subtle changes, but I'd categorize them as improvements. :)

Closes #28 
Closes #29 
Closes #32 
Closes #4
Closes #26
Closes #61

### CODE REVIEW

File: chainsync/chainsync.go [hunk 1]
```diff
@@ -2,42 +2,70 @@ package chainsync
...
-	for ts.Height >= bts.Height {
-		path = append(path, ts)
-		ts, err = c.ChainGetTipSet(ctx, types.NewTipSetKey(ts.Blocks[0].Parents...))
+	norevert := fpath[0].Type == types.HCApply
+	return norevert, nil
+}
+
+// ResolveBase returns the base TipSetKey that both left and right TipSetKey share,
+// plus a Revert/Apply set of operations to get from last to new.
+func ResolveBase(ctx context.Context, api API, left *types.TipSetKey, right types.TipSetKey) (*types.TipSetKey, []*types.TipSet, error) {
```
Comment (jsign) [inline chainsync/chainsync.go]: This methods leverates the PR I made in Lotus.
It takes a `left` and `right` tipsets, and resolves the most recent common Tipset that are in both chains.
Apart from returning that value, it also returns a slice of the path that reverts from `left` to this base tipset, and then all the path from that base tipset to `to`.

This is used by indexers which ask for `chainstore` for the last saved state in the same fork of the new discovered head, and then can use this method to get the path from which they should update since that state up to the new head.

File: cmd/server/main.go [hunk 2]
```diff
@@ -21,23 +22,35 @@ var (
...
+	repoPath, err := os.UserHomeDir()
+	if err != nil {
+		log.Errorf("error getting home dir: %s", err)
+		os.Exit(-1)
 	}
 	conf := server.Config{
 		LotusAddress:    lotusAddr,
 		LotusAuthToken:  token,
 		GrpcHostAddress: grpcAddr,
+		RepoPath:        filepath.Join(repoPath, ".texfc"),
```
Comment (jsign) [inline cmd/server/main.go]: For the moment the default repo path is `~/.texfc`, kind of arbitrary. As ToDo mentions, latter could be flagged.

File: docker/docker-compose.yaml [hunk 3]
```diff
@@ -36,8 +36,10 @@ services:
       - /var/run:/var/run:rw
       - /sys:/sys:ro
       - /var/lib/docker/:/var/lib/docker:ro
+    command:
+       - '--port=8082'
     ports:
-      - 8080:8080
+      - 8082:8082
```
Comment (jsign) [inline docker/docker-compose.yaml]: Fixed since `8080` was used by a previous PR change.

File: docker/grafana/provisioning/dashboards/TextileFC.json [hunk 4]
```diff
@@ -516,7 +516,7 @@
       "pluginVersion": "6.4.4",
```
Comment (jsign) [inline docker/grafana/provisioning/dashboards/TextileFC.json]: Mostly Grafana dashboard fixes and improvements.

File: fchost/fchost.go [hunk 5]
```diff
@@ -17,11 +17,13 @@ var (
 	log = logging.Logger("fchost")
```
Comment (jsign) [inline fchost/fchost.go]: Mostly start documenting methods. This was done not only on this file, but in mostly all ones.

File: index/ask/ask.go [hunk 6]
```diff
@@ -2,6 +2,7 @@ package ask
```
Comment (jsign) [inline index/ask/ask.go]: Mostly:
- better names
- refactoring
- bugfixes
- persist index state using `cbor` in a real datastore.

File: index/ask/metrics.go [hunk 7]
```diff
@@ -7,32 +7,32 @@ import (
 )
```
Comment (jsign) [inline index/ask/metrics.go]: Fresh review of metrics names and descriptions.

File: index/ask/types.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package ask
```
Comment (jsign) [inline index/ask/types.go]: Try to separate `ask.go` in more parts to stop growing too much.

File: index/miner/meta.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
+package miner
```
Comment (jsign) [inline index/miner/meta.go]: This new file is an extraction of code from `miners.go` which contained all the code for the miners index. 
The `meta.go` file has all the code related to metadata gathered of miners, such as location, user-agent or any off-chain data that might exist in the future.

Not only is now in a different file, but the data was separated from on-chain data too. It's now also mantained in a separate goroutine. This was done since it didn't make sense to tie off-chain metadata scrapping with on-chain data. On-chain data needs to be refreshed when the chain has a new Tipset. Off-chain data might have its own triggers that might be completely different than the chain.

TL;DR: more separation and structure.

File: index/miner/metrics.go [hunk 10]
```diff
@@ -7,41 +7,41 @@ import (
 )
```
Comment (jsign) [inline index/miner/metrics.go]: Revisited metrics names and descriptions.

File: index/miner/miner.go [hunk 11]
```diff
@@ -2,37 +2,34 @@ package miner
 
 import (
```
Comment (jsign) [inline index/miner/miner.go]: - As mentioned a bit before, this file was partitioned into: `miner.go`, `meta.go`, and `onchain.go`.
`miner.go` has some central coordination code for the index, and on-chain data gathering lives on `onchain.go` and off-chain data gathering lives on `meta.go`.
- Index information is saved/loaded in datastore
- refactors/fixes

File: index/miner/onchain.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
+package miner
```
Comment (jsign) [inline index/miner/onchain.go]: The on-chain data gathering from miners index.

File: index/miner/onchain.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+// updateOnChainIndex updates on-chain index information when a new tipset
+func (mi *MinerIndex) updateOnChainIndex(new *types.TipSet) error {
+	log.Info("updating on-chain index...")
+	var chainIndex ChainIndex
+	newtsk := types.NewTipSetKey(new.Cids...)
+	ts, err := mi.store.LoadAndPrune(mi.ctx, newtsk, &chainIndex)
+	if err != nil {
+		return err
+	}
```
Comment (jsign) [inline index/miner/onchain.go]: Here we start leveraging the new `chainstore`.
Whenever a new Tipset is discovered from the chain, `updateOnChainIndex` is called with that as a parameter.

Calling `LoadAndPrune` is like saying: _Hey, give me the last saved state that is compatible with this new tipset and is on the same fork_. With that saved state, we know that we should re-apply changes from that point towards `new`.

File: index/miner/onchain.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+		}
+	}
+	chainIndex.LastUpdated = new.Height
+	stats.Record(mctx, mRefreshDuration.M(int64(time.Since(start).Milliseconds())))
+
+	mi.lock.Lock()
+	mi.index.Chain = chainIndex
+	mi.lock.Unlock()
+
+	if err := mi.store.Save(mi.ctx, types.NewTipSetKey(new.Cids...), chainIndex); err != nil {
```
Comment (jsign) [inline index/miner/onchain.go]: Now that we have the new `chainIndex` which got built from the last saved state, up to the new Tipset. we call `Save` which basically saves this new state that corresponds to the `new` Tipset mark.

On future updates that doesn't fork from `new`, the `LoadAndPrune` call will return this state if can be used to move forward. And the cycle continues.

File: index/miner/types.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+	cbor.RegisterCborType(Power{})
+	cbor.RegisterCborType(MetaIndex{})
+	cbor.RegisterCborType(Meta{})
+	cbor.RegisterCborType(Location{})
+}
+
+// Index contains on-chain and off-chain information about miners
+type Index struct {
+	Meta  MetaIndex
+	Chain ChainIndex
```
Comment (jsign) [inline index/miner/types.go]: Now the Index has separated metadata and on-chain information.
This because those different informations are mantained by two different subcomponents (`meta.go` and `onchain.go`), so the shouldn't use locks or care about concurrency.

File: index/slashing/metrics.go [hunk 16]
```diff
@@ -6,24 +6,24 @@ import (
 )
```
Comment (jsign) [inline index/slashing/metrics.go]: Names an descriptions revisited.

File: index/slashing/slashing.go [hunk 17]
```diff
@@ -9,30 +9,37 @@ import (
 	"github.com/ipfs/go-cid"
```
Comment (jsign) [inline index/slashing/slashing.go]: Similar as others:
- refactored/fixes
- save/load index from datastore
- extracted types to `types.go`

File: index/slashing/slashing.go [hunk 18]
```diff
@@ -41,167 +48,186 @@ type SlashingIndex struct {
...
+			j = len(path)
+		}
+		if err := updateFromPath(s.ctx, s.api, &index, path[i:j]); err != nil {
+			return err
+		}
+		if err := s.store.Save(s.ctx, types.NewTipSetKey(path[j-1].Cids...), index); err != nil {
+			return err
+		}
+		stats.Record(mctx, mRefreshProgress.M(float64(i)/float64(len(path))))
 	}
```
Comment (jsign) [inline index/slashing/slashing.go]: Here's the most powerful leverage of `chainstore` and `chainsync`.
`chainstore` is use similar as said before: get last same-fork saved state from newly discovered Tipset.
Having this saved state as a starting point, we can use `ResolveBase`, which will return us a path from this base tipset state to `new`. (This is a simplified case since we know that saved state is from the same fork, so the path returned doesn't have any `revert` actions, only a fast-forward path to `new`).

Since the slashing data is on a single `int` value on chain and might change for every new Tipset, we should walk the path and start registering when this value changes for each miner. Here I use a concept of `batchSize`, mostly to save calculated state every 20 Tipsets. If for a reason the process is killed, it only lost the some of the work done.

File: index/slashing/types.go [hunk 19]
```diff
@@ -0,0 +1,19 @@
+package slashing
```
Comment (jsign) [inline index/slashing/types.go]: Types extracted here with some renaming.

File: lotus/client.go [hunk 20]
```diff
@@ -211,6 +213,20 @@ func (a *API) ChainGetTipSetByHeight(ctx context.Context, height uint64, ts *typ
 	}
 	return ts, nil
 }
+func (a *API) ChainGetPath(ctx context.Context, from types.TipSetKey, to types.TipSetKey) ([]*types.HeadChange, error) {
```
Comment (jsign) [inline lotus/client.go]: This is the new API added by us on Lotus

File: reputation/internal/source/source.go [hunk 21]
```diff
@@ -7,14 +7,16 @@ import (
 	ma "github.com/multiformats/go-multiaddr"
 )
 
+// Source is an external source of reputation information
```
Comment (jsign) [inline reputation/internal/source/source.go]: More linting improvements

File: signaler/signaler.go [hunk 22]
```diff
@@ -54,3 +60,13 @@ func (s *Signaler) Signal() {
 		}
 	}
 }
+
+// Close closes the Signaler. Any channel that wasn't explicitely unregistered,
+// is closed.
+func (s *Signaler) Close() {
```
Comment (jsign) [inline signaler/signaler.go]: A `Close` was missing to, just in case, close all existing channels and let any remaining clients now that nothing more will be pushed.

File: txndstransform/txndstransform.go [hunk 23]
```diff
@@ -0,0 +1,209 @@
+package txndstransform
```
Comment (jsign) [inline txndstransform/txndstransform.go]: This is a utility that I did in `go-threads`. It allows to Wrap a `TxnDatastore` with a prefix-namespace.
The existing one by PL only works with `datastore.Datastore` (not `TxnDatastore`).
Since we use badger, and I needed to use it as `TxnDatastore` (since I needed transactions) and also wanted to prefix each index with their own namespace... I just re-leveraged this work.

File: chainstore/chainstore.go [hunk 24]
```diff
@@ -0,0 +1,230 @@
+package chainstore
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply chainstore/chainstore.go]: Nice, good summary 💪

File: index/miner/onchain.go (hunk 13 above)
Comment (sanderpick) [reply index/miner/onchain.go]: Kickass.

File: txndstransform/txndstransform.go (hunk 23 above)
Comment (sanderpick) [reply txndstransform/txndstransform.go]: Nice! Maybe this could be it's own lib at some point.
